File: SevensGame.cs
using System;
using System.Collections.Generic;
using System.Linq;
using static MyGlobal;

static class MyGlobal{
	//全自動モード
	public const bool AUTO_MODE=false;
	//プレイヤー人数
	public const int PLAYER_NUMBER=4;
	//パス回数
	public const int PASSES_NUMBER=3;
}

//トランプカードクラス
class TrumpCard{
	public static readonly string[] suitStrs={"▲","▼","◆","■","Jo","JO"};
	public static readonly string[] powerStrs={"Ａ","２","３","４","５","６","７","８","９","10","Ｊ","Ｑ","Ｋ","KR"};
	public const int suits=4;
	public const int powers=13;
	public readonly string name;
	public readonly int power;
	public readonly int suit;
	public TrumpCard(int suit,int power){
		this.name=suitStrs[suit]+powerStrs[power];
		this.power=power;
		this.suit=suit;
	}
}

//トランプの束クラス
class TrumpDeck{
	public static readonly Random rand=new Random();
	readonly IEnumerator<TrumpCard> g;
	readonly List<TrumpCard> deck=new List<TrumpCard>();

	IEnumerator<TrumpCard> trumpIter(List<TrumpCard> deck){
		foreach(var v in deck){
			yield return v;
		}
	}
		
	public int count{get{return deck.Count;}}

	public TrumpDeck(){
		for(var suit=0;suit<TrumpCard.suits;suit++){
			for(var power=0;power<TrumpCard.powers;power++){
				deck.Add(new TrumpCard(suit,power));
			}
		}

		/* Joker
		deck.Add(new TrumpCard(4,TrumpCard.powers));
		deck.Add(new TrumpCard(5,TrumpCard.powers));
		*/

		g=trumpIter(deck);
	}

	public void shuffle(){
		for(var i=0;i<deck.Count-1;i++){
			var r=rand.Next(i,deck.Count);
			var tmp=deck[i];
			deck[i]=deck[r];
			deck[r]=tmp;
		}
	}

	public TrumpCard draw(){
		g.MoveNext();
		return g.Current;
	}
}

//プレイヤークラス
class Player{
	public readonly List<TrumpCard> deck=new List<TrumpCard>();
	public int id;
	public string name;
	public bool isGameOut;

	public Player(int id,string name){
		this.id=id;
		this.name=name;
	}

	public static void sortRefDeck(List<TrumpCard> deck){
		Func<TrumpCard,int> sortValue=v=>v.suit*TrumpCard.powers+v.power;
		deck.Sort((a,b)=>sortValue(a)-sortValue(b));
	}

	public void sortDeck(){sortRefDeck(deck);}

	public void addCard(TrumpCard card){
		deck.Add(card);
	}

	public void removeCard(string cardName){
		deck.Remove(deck.Find(v=>v.name==cardName));
	}

	public int existCard(string cardName){
		return deck.FindIndex(v=>v.name==cardName);
	}

	public void gameOut(){
		isGameOut=true;
	}
}

//トランプの場クラス
class TrumpField{
	protected List<Player> players;
	public readonly List<TrumpCard> deck=new List<TrumpCard>();
	public void sortDeck(){Player.sortRefDeck(deck);}

	public TrumpField(List<Player> players){
		this.players=players;
	}

	public virtual void useCard(Player player,TrumpCard card){
		deck.Add(card);
		player.removeCard(card.name);
	}

	public virtual void view(){
		Console.WriteLine(String.Join(" ",deck.Select(v=>v.name)));
	}
}

//七並べの列クラス
class SevensLine{
	const int sevenIndex=6;
	public readonly bool[] cardLine=new bool[TrumpCard.powers];

	public SevensLine(){
		cardLine[sevenIndex]=true;
	}

	public int rangeMin(){
		int i;
		for(i=sevenIndex;0<=i;i--){
			if(!cardLine[i]) return i;
		}
		return i;
	}

	public int rangeMax(){
		int i;
		for(i=sevenIndex;i<TrumpCard.powers;i++){
			if(!cardLine[i]) return i;
		}
		return i;
	}

	public bool checkUseCard(int power){
		if(
			power==TrumpCard.powers ||
			power==rangeMin() ||
			power==rangeMax()
		) return true;
		return false;
	}

	public void useCard(int power){
		cardLine[power]=true;
	}
}

//七並べクラス 
class Sevens:TrumpField{
	const int tenhoh=0xFF;
	public readonly SevensLine[] lines=Enumerable.Range(0,TrumpCard.suits).Select(x=>new SevensLine()).ToArray();
	readonly int[] rank;
	public int clearCount;
	
	public Sevens(List<Player> players):base(players){
		rank=new int[this.players.Count];
		clearCount=0;

		for(var i=0;i<TrumpCard.suits;i++){
			var cardSevenName=TrumpCard.suitStrs[i]+TrumpCard.powerStrs[6];
			for(var n=0;n<this.players.Count;n++){
				var p=this.players[n];
				var cardSevenIndex=p.existCard(cardSevenName);
				if(-1<cardSevenIndex){
					var card=p.deck[cardSevenIndex];
					Console.WriteLine($"{p.name} が{card.name}を置きました。");
					useCard(p,card);
					if(p.deck.Count==0){
						Console.WriteLine($"{p.name} 【-- 天和 --】\n");
						rank[n]=tenhoh;
						p.gameOut();
					}
					break;
				}
			}
		}
		Console.WriteLine();
	}

	public override void useCard(Player player,TrumpCard card){
		lines[card.suit].useCard(card.power);
		base.useCard(player,card);
	}

	public bool checkUseCard(TrumpCard card){
		return lines[card.suit].checkUseCard(card.power);
	}

	public bool tryUseCard(Player player,TrumpCard card){
		if(!checkUseCard(card)) return false;
		useCard(player,card);
		return true;
	}

	public bool checkPlayNext(Player player,int passes){
		if(0<passes) return true;
		foreach(var card in player.deck){
			if(checkUseCard(card)){
				return true;
			}
		}
		return false;
	}

	public void gameClear(Player player){
		clearCount++;
		rank[player.id]=clearCount;
		player.gameOut();
	}

	public void gameOver(Player player){
		rank[player.id]=-1;
		for(var i=player.deck.Count-1;i>=0;i--){
			useCard(player,player.deck[i]);
		}
		player.gameOut();
	}

	public bool checkGameEnd(){
		foreach(var v in rank){
			if(v==0) return false;
		}
		return true;
	}

	public override void view(){
		var s="";
		for(var i=0;i<TrumpCard.suits;i++){
			var ss="";
			for(var n=0;n<TrumpCard.powers;n++){
				if(lines[i].cardLine[n]){
					s+=TrumpCard.suitStrs[i];
					ss+=TrumpCard.powerStrs[n];
				}
				else{
					s+="◇";
					ss+="◇";
				}
			}
			s+="\n"+ss+"\n";
		}
		Console.WriteLine(s);
	}

	public void result(){
		Console.WriteLine("\n【Game Result】");
		string rankStr;
		for(var i=0;i<rank.Length;i++){
			if(rank[i]==tenhoh){
				rankStr="天和";
			}
			else if(0<rank[i]){
				rankStr=$"{rank[i]}位";
			}
			else{
				rankStr="GameOver...";
			}
			Console.WriteLine($"{players[i].name}: {rankStr}");
		}
	}
}

//カーソル選択モジュール
static class SelectCursors{
	public static int SelectCursor(List<string> items){
		var cursor=0;
		//カーソルの移動
		Action<int,int> move=(x,max)=>{
			cursor+=x;
			if(cursor<0) cursor=0;
			if(max-1<cursor) cursor=max-1;
		};

		//カーソルの表示
		Action view=()=>{
			var select=new bool[items.Count];
			select[cursor]=true;
			var s="";
			for(int i=0;i<items.Count;i++){
				s+=select[i]? $"[{items[i]}]": items[i];
			}
			Console.Write($"{s}\r");
		};

		view();
		for(;;){
			var ch=Console.ReadKey(true);
			if(ch.Key==ConsoleKey.Enter){
				Console.WriteLine();
				return cursor;
			}
			if(ch.Key==ConsoleKey.LeftArrow) move(-1,items.Count);	//左
			if(ch.Key==ConsoleKey.RightArrow) move(1,items.Count);	//右
			view();
		}
	}
}

//七並べプレイヤークラス
class SevensPlayer:Player{
	protected int passes;
	public SevensPlayer(int id,string name,int passes):base(id,name){
		this.passes=passes;
	}

	public virtual void selectCard(Sevens field){
		if(isGameOut) return;
		if(!field.checkPlayNext(this,passes)){
			field.gameOver(this);
			field.view();
			Console.WriteLine($"{name} GameOver...\n");
			return;
		}

		Console.WriteLine($"【{name}】Cards: {deck.Count} Pass: {passes}");
		var items=new List<string>(deck.Select(v=>v.name));
		if(0<passes) items.Add("PS:"+passes);

		for(;;){
			var cursor=SelectCursors.SelectCursor(items);

			if(0<passes && items.Count-1==cursor){
				passes--;
				field.view();
				Console.WriteLine($"残りパスは{passes}回です。\n");
				break;
			}
			else if(field.tryUseCard(this,deck[cursor])){
				field.view();
				Console.WriteLine($"俺の切り札!! >「{items[cursor]}」\n");
				if(deck.Count==0){
					Console.WriteLine($"{name} Congratulations!!\n");
					field.gameClear(this);
				}
				break;
			}
			else{
				Console.WriteLine("そのカードは出せないのじゃ…\n");
				continue;
			}
		}
	}
}

//七並べAIプレイヤークラス
class SevensAIPlayer:SevensPlayer{
	public SevensAIPlayer(int id,string name,int passes):base(id,name,passes){}

	public override void selectCard(Sevens field){
		if(isGameOut) return;
		if(!field.checkPlayNext(this,passes)){
			field.gameOver(this);
			field.view();
			Console.WriteLine($"{name}> もうだめ...\n");
			return;
		}

		Console.WriteLine($"【{name}】Cards: {deck.Count} Pass: {passes}");
		var items=new List<string>(deck.Select(v=>v.name));
		if(0<passes) items.Add("PS:"+passes);

		Console.Write("考え中...\r");
		System.Threading.Thread.Sleep(1000);

		var passCharge=0;

		for(;;){
			var cursor=TrumpDeck.rand.Next(items.Count);

			if(0<passes && items.Count-1==cursor){
				if(passCharge<3){
					passCharge++;
					continue;
				}
				passes--;
				Console.WriteLine($"パスー (残り{passes}回)\n");
				break;
			}
			else if(field.tryUseCard(this,deck[cursor])){
				Console.WriteLine($"これでも食らいなっ >「{items[cursor]}」\n");
				if(deck.Count==0){
					Console.WriteLine($"{name}> おっさき～\n");
					field.gameClear(this);
				}
				break;
			}
			else continue;
		}
	}
}

//メイン処理
class Program{
	static void Main(){
		for(var i=0;i<100;i++){
			Console.WriteLine();
		}

		Console.WriteLine(@"
/---------------------------------------/
/                 七並べ                /
/---------------------------------------/

");

		var trp=new TrumpDeck();
		trp.shuffle();

		var p=new List<SevensPlayer>();
		var pid=0;
		if(!AUTO_MODE){
			p.Add(new SevensPlayer(pid,"Player",PASSES_NUMBER));
			pid++;
		}

		for(var i=0;i<PLAYER_NUMBER-(AUTO_MODE?0:1);i++){
			p.Add(new SevensAIPlayer(pid,$"CPU {i+1}",PASSES_NUMBER));
			pid++;
		}

		for(var i=0;i<trp.count;i++){
			p[i%PLAYER_NUMBER].addCard(trp.draw());
		}

		foreach(var v in p){
			v.sortDeck();
		}

		var field=new Sevens(p.Select(v=>(Player)v).ToList());

		for(;;){
			field.view();
			foreach(var v in p){
				v.selectCard(field);
				if(field.checkGameEnd()) goto selectLoop;
			}
		}
		selectLoop:

		field.view();
		field.result();
		Console.ReadLine();
	}
}

File: SevensGame.vb
Option Strict On
Imports System.Collections.Generic
Imports System.Linq
Imports myGlobal

Module myGlobal
	'全自動モード
	Public Const AUTO_MODE=False
	'プレイヤー人数
	Public Const PLAYER_NUMBER=4
	'パス回数
	Public Const PASSES_NUMBER=3
End Module

'トランプカードクラス
Class TrumpCard
	Public Shared ReadOnly suitStrs As String()={"▲","▼","◆","■","Jo","JO"}
	Public Shared ReadOnly powerStrs As String()={"Ａ","２","３","４","５","６","７","８","９","10","Ｊ","Ｑ","Ｋ","KR"}
	Public Const suits As Integer=4
	Public Const powers As Integer=13
	Public ReadOnly name As String
	Public ReadOnly power As Integer
	Public ReadOnly suit As Integer
	Sub New(suit As Integer,power As Integer)
		Me.name=suitStrs(suit) & powerStrs(power)
		Me.power=power
		Me.suit=suit
	End Sub
End Class

'トランプの束クラス
Class TrumpDeck
	Public Shared ReadOnly rand As New Random()
	ReadOnly g As IEnumerator(Of TrumpCard)
	ReadOnly deck As New List(Of TrumpCard)()

	Private Iterator Function trumpIter(deck As List(Of TrumpCard)) As IEnumerator(Of TrumpCard)
		For Each v In deck
			Yield v
		Next
	End Function

	ReadOnly Property count() As Integer
		Get
			return deck.Count:End Get:End Property

	Sub New()
		For suit=0 To TrumpCard.suits-1
			For power=0 To TrumpCard.powers-1
				deck.Add(New TrumpCard(suit,power))
			Next
		Next

		' Joker
		'deck.Add(New TrumpCard(4,TrumpCard.powers))
		'deck.Add(New TrumpCard(5,TrumpCard.powers))

		g=trumpIter(deck)
	End Sub

	Sub shuffle()
		For i=0 To deck.Count-2
			Dim r=rand.Next(i,deck.Count)
			Dim tmp=deck(i)
			deck(i)=deck(r)
			deck(r)=tmp
		Next
	End Sub

	Function draw() As TrumpCard
		g.MoveNext()
		Return g.Current
	End Function
End Class

'プレイヤークラス
Class Player
	Public ReadOnly deck As New List(Of TrumpCard)()
	public  name As String
	Public isGameOut As Boolean

	Sub New(name As String)
		Me.name=name
	End Sub

	Shared Sub sortRefDeck(deck As List(Of TrumpCard))
		Dim sortValue As Func(Of TrumpCard,Integer)=Function(v) v.suit*TrumpCard.powers+v.power
		deck.Sort(Function(a,b) sortValue(a)-sortValue(b))
	End Sub
	
	Sub sortDeck
		Player.sortRefDeck(deck):End Sub

	Sub addCard(card As TrumpCard)
		deck.Add(card)
	End Sub

	Sub removeCard(cardName As String)
		deck.Remove(deck.Find(Function(v) v.name=cardName))
	End Sub

	Function existCard(cardName As String) As Integer
		Return deck.FindIndex(Function(v) v.name=cardName)
	End Function

	Sub gameOut()
		isGameOut=True
	End Sub
End Class

'カーソル選択モジュール
Module SelectCursors
	Function SelectCursor(items As List(Of String)) As Integer
		Dim cursor=0
		'カーソルの移動
		Dim move As Action(Of Integer,Integer)=Sub(x,max)
			cursor+=x
			If cursor<0 Then cursor=0
			If max-1<cursor Then cursor=max-1
		End Sub

		'カーソルの表示
		Dim view As Action=Sub()
			Dim _select(items.Count-1) As Boolean
			_select(cursor)=True
			Dim s=""
			For  i=0 To items.Count-1
				s+=If (_select(i),$"[{items(i)}]",items(i))
			Next
			Console.Write($"{s}{vbCr}")
		End Sub

		view()
		Do
			Dim ch=Console.ReadKey(True)
			If  ch.Key=ConsoleKey.Enter Then
				Console.WriteLine()
				Return cursor
			End If 
			If  ch.Key=ConsoleKey.LeftArrow Then move(-1,items.Count)	'左
			If  ch.Key=ConsoleKey.RightArrow Then move(1,items.Count)	'右
			view()
		Loop
	End Function
End Module

'七並べプレイヤークラス
Class SevensPlayer
	Inherits Player
	Public passes As Integer
	Sub New(name As String,passes As Integer)
		MyBase.New(name)
		Me.passes=passes
	End Sub

	OverRidable Sub selectCard(field As Sevens,index As Integer)
		If isGameOut Then Exit Sub
		If Not field.checkPlayNext(Me) Then
			field.gameOver(Me,index)
			field.view()
			Console.WriteLine($"{name} GameOver...{vbLf}")
			Exit Sub
		End If

		Console.WriteLine($"【{name}】Cards: {deck.Count} Pass: {passes}")
		Dim items=New List(Of string)(deck.Select(Function(v) v.name))
		If 0<passes Then items.Add("PS:" & passes)

		Do
			Dim cursor=SelectCursors.SelectCursor(items)

			If 0<passes And items.Count-1=cursor Then
				passes-=1
				field.view()
				Console.WriteLine($"残りパスは{passes}回です。{vbLf}")
				Exit Do

			ElseIf field.tryUseCard(Me,deck(cursor)) Then
				field.view()
				Console.WriteLine($"俺の切り札!! >「{items(cursor)}」{vbLf}")
				If deck.Count=0 Then
					Console.WriteLine($"{name} Congratulations!!{vbLf}")
					field.gameClear(Me,index)
				End If
				Exit Do

			Else
				Console.WriteLine($"そのカードは出せないのじゃ…{vbLf}")
				Continue Do
			End If
		Loop
	End Sub
End Class

'七並べAIプレイヤークラス
Class SevensAIPlayer
	Inherits SevensPlayer
	Sub New(name As String,passes As Integer)
		MyBase.New(name,passes)
	End Sub

	Overrides Sub selectCard(field As Sevens,index As Integer)
		If isGameOut Then Exit Sub
		If Not field.checkPlayNext(Me) Then
			field.gameOver(Me,index)
			field.view()
			Console.WriteLine($"{name}> もうだめ...{vbLf}")
			Exit Sub
		End If

		Console.WriteLine($"【{name}】Cards: {deck.Count} Pass: {passes}")
		Dim items=new List(Of string)(deck.Select(Function(v) v.name))
		If 0<passes Then items.Add("PS:" & passes)

		Console.Write($"考え中...{vbCr}")
		Threading.Thread.Sleep(1000)

		Dim passCharge=0

		Do
			Dim cursor=TrumpDeck.rand.Next(items.Count)

			If 0<passes And items.Count-1=cursor Then
				If passCharge<3 Then
					passCharge+=1
					Continue Do
				End If

				passes-=1
				Console.WriteLine($"パスー (残り{passes}回){vbLf}")
				Exit Do

			ElseIf field.tryUseCard(Me,deck(cursor)) Then
				Console.WriteLine($"これでも食らいなっ >「{items(cursor)}」{vbLf}")
				If deck.Count=0 Then
					Console.WriteLine($"{name}> おっさき～{vbLf}")
					field.gameClear(Me,index)
				End If
				Exit Do

			Else
				Continue Do
			End If
		Loop
	End Sub
End Class

'トランプの場クラス
Class TrumpField
	Public ReadOnly deck As New List(Of TrumpCard)()

	Sub sortDeck()
		Player.sortRefDeck(deck):End Sub

	OverRidable Sub useCard(player As SevensPlayer,card As TrumpCard)
		deck.Add(card)
		player.removeCard(card.name)
	End Sub

	OverRidable Sub view()
		Console.WriteLine(String.Join(" ",deck.Select(Function(v) v.name)))
	End Sub
End Class

'七並べの列クラス
Class SevensLine
	Const sevenIndex=6
	Public ReadOnly cardLine(TrumpCard.powers-1) As Boolean

	Sub New()
		cardLine(sevenIndex)=True
	End Sub

	Function rangeMin() As Integer
		Dim i As Integer
		For i=sevenIndex To 0 Step -1
			If Not cardLine(i) Then Return i
		Next
		Return i
	End Function

	Function rangeMax() As Integer
		Dim i As Integer
		For i=sevenIndex To TrumpCard.powers-1
			If Not cardLine(i) Then Return i
		Next
		Return i
	End Function

	Function checkUseCard(power As Integer) As Boolean
		Select Case power
			Case TrumpCard.powers,rangeMin(),rangeMax()
				Return True
			Case Else
				Return False
		End Select
	End Function

	Sub useCard(power As Integer)
		cardLine(power)=True
	End Sub
End Class

'七並べクラス 
Class Sevens
	Inherits TrumpField
	Const tenhoh=&HFF
	Public ReadOnly lines As SevensLine()
	ReadOnly rank As Integer()
	Public clearCount As Integer

	Sub New(players As List(Of SevensPlayer))
		MyBase.New()
		lines=Enumerable.Range(0,TrumpCard.suits).Select(Function(x) New SevensLine()).ToArray()
		ReDim rank(players.Count-1)
		clearCount=0

		For i=0 To TrumpCard.suits-1
			Dim cardSevenName=TrumpCard.suitStrs(i) & TrumpCard.powerStrs(6)
			For n=0 To players.Count-1
				Dim p=players(n)
				Dim cardSevenIndex=p.existCard(cardSevenName)
				If -1<cardSevenIndex Then
					Dim card=p.deck(cardSevenIndex)
					Console.WriteLine($"{p.name} が{card.name}を置きました。")
					useCard(p,card)
					If p.deck.Count=0 Then
						Console.WriteLine($"{p.name} 【-- 天和 --】{vbLf}")
						rank(n)=tenhoh
						p.gameOut()
					End If
					Exit For
				End If
			Next
		Next
		Console.WriteLine()
	End Sub

	Overrides Sub useCard(player As SevensPlayer,card As TrumpCard)
		lines(card.suit).useCard(card.power)
		MyBase.useCard(player,card)
	End Sub

	Function checkUseCard(card As TrumpCard) As Boolean
		Return lines(card.suit).checkUseCard(card.power)
	End Function

	Function tryUseCard(player As SevensPlayer,card As TrumpCard) As Boolean
		If Not checkUseCard(card) Then Return False
		useCard(player,card)
		Return True
	End Function

	Function checkPlayNext(player As SevensPlayer) As Boolean
		If 0<player.passes Then Return True
		For Each card In player.deck
			If checkUseCard(card) Then _
				Return True
		Next
		Return False
	End Function

	Sub gameClear(player As SevensPlayer,index As Integer)
		clearCount+=1
		rank(index)=clearCount
		player.gameOut()
	End Sub

	Sub gameOver(player As SevensPlayer,index As Integer)
		rank(index)=-1
		For i=player.deck.Count-1 To 0 Step -1
			useCard(player,player.deck(i))
		Next
		player.gameOut()
	End Sub

	Function checkGameEnd() As Boolean
		For Each v In rank
			If v=0 Then Return False
		Next
		Return True
	End Function

	OverRides Sub view()
		Dim s=""
		For i=0 To TrumpCard.suits-1
			Dim ss=""
			For n=0 To TrumpCard.powers-1
				If lines(i).cardLine(n) Then
					s &=TrumpCard.suitStrs(i)
					ss &=TrumpCard.powerStrs(n)
				Else
					s &="◇"
					ss &="◇"
				End If
			Next
			s &=vbLf & ss & vbLf
		Next
		Console.WriteLine(s)
	End Sub

	Sub result(players As List(Of SevensPlayer))
		Console.WriteLine($"{vbLf}【Game Result】")
		Dim rankStr As String
		For i=0 To rank.Length-1
			If rank(i)=tenhoh Then
				rankStr="天和"
			ElseIf 0<rank(i) Then
				rankStr=$"{rank(i)}位"
			Else
				rankStr="GameOver..."
			End If
			Console.WriteLine($"{players(i).name}: {rankStr}")
		Next
	End Sub
End Class

'メイン処理
Module Program
	Sub Main()
		For i=1 To 100
			Console.WriteLine()
		Next

Console.WriteLine(
$"/---------------------------------------/
/                 七並べ                /
/---------------------------------------/

")
		Dim trp=New TrumpDeck()
		trp.shuffle()

		Dim p=new List(Of SevensPlayer)()

		If Not AUTO_MODE Then
			p.Add(New SevensPlayer("Player",PASSES_NUMBER))
		End If

		For i=1 To PLAYER_NUMBER-(If(AUTO_MODE,0,1))
			p.Add(New SevensAIPlayer($"CPU {i}",PASSES_NUMBER))
		Next

		For i=0 To trp.count-1
			p(i Mod PLAYER_NUMBER).addCard(trp.draw())
		Next

		For Each v In p
			v.sortDeck()
		Next

		Dim field=New Sevens(p)

		Do
			field.view()
			For i=0 To p.Count-1
				p(i).selectCard(field,i)
				If field.checkGameEnd() Then Exit Do
			Next
		Loop

		field.view()
		field.result(p)
		Console.ReadLine()
	End Sub
End Module
